### DISCUSSION TITLE
Just want to say thanks!

### DISCUSSION THREAD

**Discussion by FPVBadGuy** [SammyIAm/Moppy2#125] [General]
Hi Sam! 

I just signed up to Github just to get notifications on your sweet project! I came across Moppy last winter. It was an amazing experience (but also was looking a bit chaotic to be honest).

Me and a friend of mine are working on a Youtube-Video for a big German Electronics warehouse to introduce the Moppy project to a more or less big audience. For this we are integrating 8 drives, an Arduino UNO and a cheap power supply into a cheap IKEA table (wich us luck). 

What I've always missed is some kind of tutorial on making HDD or solenloid drums work using a shift register. I've looked through the "issues"-tab, but couldn't figure it out. It's not for the video we are producing, but I would like to continue the project off camera. Do you have a helpful source of information on how to wire up a HDD?

I would love to see this project grow a lot in the future, it is a cool way to spend time in the shed 👍 

A big thumbs up from Germany! Stay healthy and keep buzzing out some fine music!

**Comment by samphonic** [comment]
Hey there!

Sorry it's taken me so long to get back to you, life's been really busy and as much as I enjoy musical floppy drives, they're not quite priority number one, you know? 😉 

Your project sounds awesome!  I love seeing other implementations, and especially some of the really nicely polished ones (I'm also a fan of doing cool stuff with IKEA furniture, so it's a win-win).  I'm not aware, off hand, of any tutorials on HDD or solenoid drums using Moppy.  I know there are some GitHub users who have done that in the past, so you **might** be able to find something if they've put anything in their own forks of the project, but probably nothing full-featured.

I have wired up some solenoids myself though to ring handbells (I burnt out some of the circuits though and never got it finished enough to film... still sitting in my garage waiting to be revived), so I can maybe point you at least in the general direction:

There's a `ShiftRegister` instrument included in the current codebase that you can switch to by changing [the appropriate line in MoppyConfig.h](https://github.com/Sammy1Am/Moppy2/blob/master/Microcontroller/Moppy2-Arduino/src/MoppyConfig.h#L20).  [The implementation in ShiftRegister.cpp](https://github.com/Sammy1Am/Moppy2/blob/master/Microcontroller/Moppy2-Arduino/src/MoppyInstruments/ShiftRegister.cpp) has a few comments and the pin assignments that may help a little bit with connecting the Arduino to a shift-register.  Here's loosely what you'd do:

- Wire shift register(s) to the Arduino.
- Update `NUM_NOTES` and `SHIFT_DATA_BYTES` in `ShiftRegister.cpp` to match the number and size of your shift registers.
- You'll need to connect the logic-level output of the shit registers to something that can handle higher currents.  If you're not going to be switching many times per second then relays are a good option (I've used [modules like this one](https://www.aliexpress.com/item/32811948277.html), though there are a lot of similar options everywhere).
- Connect up a power supply to the inputs for the relays, and the solenoids, HDDs, etc. to the outputs.

MIDI drums are traditionally on channel 10 (though they can be anywhere), and [each different drum is a different "note"](https://soundprogramming.net/file-formats/general-midi-drum-note-numbers/).  You can either edit your MIDI files to make sure all the notes get played by the Arduino you've got set up, or you can somehow write some code to map the MIDI drum notes to the notes on the Arduino in real-time.

The velocity of the MIDI note determines how long the solenoid will be powered for, and this is adjustable[ in the code](https://github.com/Sammy1Am/Moppy2/blob/master/Microcontroller/Moppy2-Arduino/src/MoppyInstruments/ShiftRegister.cpp#L28-L29) as well.

If you're using relays and all you're connecting is the solenoid or HDD, you should be fine.  If you decide to connect anything else to the output (for example, I decided to connect LEDs that light up when the solenoid is triggered) you'll want to make sure you have a flyback diode connected to the solenoid (so, for example, you don't fry the LEDs on the very first run like I did ☹️).

<hr/>

That should hopefully at least give you an idea of where to go with things.  If you're feeling really adventurous, you could always document your process and turn it into a tutorial-- I'd be happy to link to it from this repo.  Additionally, I'm happy to help if you run into any weirdness along the way, and hopefully I'll be able to respond sooner next time.

Good luck!  Maybe share a video of your big project when it's finished!

**Reply by FPVBadGuy** [reply]
Hey Sam! 
Thanks for the great support! I will get back to moppy and try it soon. 
Btw: the video is out now. It's in German, but if you can take a look if you're interested: [https://www.youtube.com/watch?v=szLkVsqO3NI](url)

**Comment by Oldmanspub** [comment]
Hi Sam!

I've built this and successfully compiled for Java8 for run on older computer. I'm currently building a rig with 8 drives for  BRSzK exhibition of retro computer festival at Füle (Bacsis Retro Számítógép Kiállítás)  https://brszk.hu/
This will be a new old use of old equipment and would ask your approve to put your name and github link on my rig to show your resource. (The event will be free, you are also welcome to visit!)
Cheers,

**Comment by samphonic** [comment]
Hi there!

Happy to hear you were able to get it working, and I am happy for you to put my name and GitHub link on it (that way anyone who wants to try to build one for themselves can get started more easily).  I appreciate the invitation to visit, I checked out the website and it does sound like a lot of fun, but life's a bit busy for a trip to Europe at the moment 😉.

I hope the event goes well though.  Maybe come back and share a picture of your rig here when it's over :)
